crates/experimentation_client/src/lib.rs

```diff
@@ -213,37 +278,123 @@ impl Client {
+        {
+            Some(id) => id,
+            None => return Ok(None),
+        };
+
+        let experiment_id = variant_id
+            .split('-')
+            .next()
+            .and_then(|id| id.parse::<i64>().ok())
+            .ok_or("Invalid variant ID format")?;
+
+        let experiments = self.experiments.read().await;
+        let experiment = match experiments.get(&experiment_id.to_string()) {
+            Some(exp) if exp.status.active() => exp,
+            _ => return Ok(None),
+        };
+
+        Ok(experiment
+            .variants
+            .iter()
+            .find(|v| v.id == *variant_id)
+            .cloned())
+    }
+
+    pub async fn decide_variant_with_variant_ids(
+        &self,
+        experiment_group: &ExperimentGroup,
+        variant_ids: &[String],
+    ) -> Result<Option<Variant>, String> {
+        let member_experiment_ids = &experiment_group
+            .member_experiment_ids
+            .iter()
+            .cloned()
+            .collect::<HashSet<i64>>();
+
+        // Parse variant IDs and collect experiment IDs, checking for duplicates
+        let mut req_experiment_ids: HashMap<i64, &str> = HashMap::new();
+        for variant_id in variant_ids {
+            let experiment_id = variant_id
+                .split('-')
+                .next()
+                .and_then(|id| id.parse::<i64>().ok())
+                .ok_or({
+                    "Invalid variant ID format. Ensure the variant ID is in the correct format."
+                })?;
+
+            if req_experiment_ids
+                .insert(experiment_id, variant_id)
+                .is_some()
+            {
+                return Err(
+                    "Multiple variants of the same experiment_id provided.".to_string()
+                );
             }
         }
-        let variant_count = applicable_variants.len() as u8;
-        let range = (traffic * variant_count) as i32;
-        if (toss as i32) >= range {
-            return Ok(None);
+
+        // Find common experiment IDs and ensure at most one
+        let common_experiments: Vec<i64> = member_experiment_ids
+            .intersection(&req_experiment_ids.keys().cloned().collect())
+            .cloned()
+            .collect();
+
+        if common_experiments.len() > 1 {
+            return Err(
+                "Multiple variants of the same experiment group provided.".to_string()
+            );
+        }
+
+        // Process the single common experiment if it exists
+        if let Some(&exp_id) = common_experiments.first() {
+            let variant_id = req_experiment_ids[&exp_id];
+
+            // Get experiment and check if active
+            let experiments = self.experiments.read().await;
+            let experiment = match experiments.get(&exp_id.to_string()) {
+                Some(exp) if exp.status.active() => exp,
+                _ => return Ok(None),
+            };
+
+            if experiment.status.active() {
```

Comment:
this active status check is repetitive , also you are doing early return please verify it that is correct